proptools/compare.go

```diff
@@ -0,0 +1,248 @@
+		return deepCompareStructs(a, b, depth)
+	case reflect.Ptr, reflect.Interface:
+		return deepComparePointers(a, b, depth)
+	case reflect.Int, reflect.String, reflect.Bool:
+		return comparePrimitives(a, b, depth)
+	default:
+		panic(fmt.Sprintf("unrecognized type %v of value %#v\n", a.item.Kind(), a.item))
+
+	}
+}
+
+// compares two pointers in the same manner as reflect.deepEqual
+func deepComparePointers(a namedValue, b namedValue, depth int) (comparison Comparison) {
+	if a.item.IsNil() || b.item.IsNil() {
+		if a.item.IsNil() && b.item.IsNil() {
+			return equal
+		} else {
+
+			if a.item.IsNil() {
+				return different(fmt.Sprintf("%s is nil whereas %s = %#v", a.name, b.name, b.item))
+			} else {
+				return different(fmt.Sprintf("%s is nil whereas %s = %#v", b.name, a.name, a.item))
+			}
+		}
+
+	}
+	return deepCompareGeneric(namedValue{a.name, a.item.Elem()},
+		namedValue{b.name, b.item.Elem()},
+		depth)
+}
+
+// compares two non-struct items
+func comparePrimitives(actual namedValue, expected namedValue, depth int) (comparison Comparison) {
+	// TODO can we find a simpler implementation of this method that still works with private fields?
+	var a, b interface{}
+	if actual.item.Kind() != expected.item.Kind() {
+		return different(fmt.Sprintf("%s is of type %s whereas %s is of type %s", actual.name, actual.item.Kind(), expected.name, expected.item.Kind()))
+	}
+	switch actual.item.Kind() {
+	case reflect.String:
+		a = actual.item.String()
+		b = expected.item.String()
+	case reflect.Bool:
+		a = actual.item.Bool()
+		b = expected.item.Bool()
+	case reflect.Int:
+		a = actual.item.Int()
+		b = expected.item.Int()
+	default:
+		panic(fmt.Sprintf("unrecognized types, %s (%#v) and %s (%#v)", actual.name, actual.item, expected.name, expected.item))
+	}
+	if reflect.DeepEqual(a, b) {
+		return equal
+	} else {
+		return different(fmt.Sprintf("%s = %#v whereas %s = %#v", actual.name, actual.item, expected.name, expected.item))
+	}
+}
+
+// compares two maps
+func deepCompareMaps(a namedValue, b namedValue, depth int) (comparison Comparison) {
+	aItem := a.item
+	bItem := b.item
+	aKeys := aItem.MapKeys()
+	bKeys := bItem.MapKeys()
+	for _, aKey := range aKeys {
+		if !bItem.MapIndex(aKey).IsValid() {
+			aValue := aItem.MapIndex(aKey)
+			return different(fmt.Sprintf("%v contains key %#v (with value %#v) but %v does not", a.name, aKey, aValue, b.name))
+		}
+	}
+	for _, bKey := range bKeys {
+		if !aItem.MapIndex(bKey).IsValid() {
+			bValue := bItem.MapIndex(bKey)
+			return different(fmt.Sprintf("%v contains key %#v (with value %#v) but %v does not", b.name, bKey, bValue, a.name))
+		}
+	}
+	for _, key := range aKeys {
+		keyText := fmt.Sprintf("[%v(%t)]", key, key)
+
+		comparison := deepCompareGeneric(namedValue{a.name + keyText, aItem.MapIndex(key)},
+			namedValue{b.name + keyText, bItem.MapIndex(key)},
+			depth,
+		)
+		if !comparison.Equal {
+			return comparison
+		}
+
+	}
+	return equal
+}
+
+// compares two arrays
```

Comment:
compares two arrays or slices

document that this compares [0,len()), and not [0,cap())